### ISSUE TITLE
Add support for `bigint` primitive

### ISSUE THREAD

**Issue by cmumatt** [nanofuzz/nanofuzz#117]
Add support for `bigint` primitive

**Comment by ZiyaoWei** [comment]
Seems to be unsupported still - can I take a try? Could be a gentle introduction to the TS specific stuff - I mentioned before adding support for null but that seems fairly involved 😅 and maybe I can start by working on the current issue first.

**Comment by cmumatt** [comment]
Absolutely---please feel free!

**Comment by ZiyaoWei** [comment]
It seems `number` has internal support for signed vs. unsigned but it is not shown - I assume it's since users can set min = 0? If so I assume bigint doesn't need that and maybe we can even delete the setting for `number` too?

**Comment by ZiyaoWei** [comment]
Another "fun" serialization issue lol - there are a lot of round tripping using JSON5, curious are those there for a specific reason maybe normalization?

However, bigint doesn't work with JSON serialization and JSON.stringify raises an error, while JSON5.stringify... silently ignores bigint inputs exactly as no one would have expected 😅 

```
> JSON5.stringify(-45n)
undefined
> JSON.stringify(-45n)
Uncaught TypeError: Do not know how to serialize a BigInt
    at JSON.stringify (<anonymous>)
```

Would it be possible to delete the round tripping, use a different serialization library, or do you have any other suggestions?

**Comment by cmumatt** [comment]
> It seems `number` has internal support for signed vs. unsigned but it is not shown - I assume it's since users can set min = 0? If so I assume bigint doesn't need that and maybe we can even delete the setting for `number` too?

Yes, signed vs. unsigned is a very old setting. I believe we only used it at the very beginning prior to adding ranges, so agree that it's superfluous at this point.

**Comment by cmumatt** [comment]
> Another "fun" serialization issue lol - there are a lot of round tripping using JSON5, curious are those there for a specific reason maybe normalization?
> 
> However, bigint doesn't work with JSON serialization and JSON.stringify raises an error, while JSON5.stringify... silently ignores bigint inputs exactly as no one would have expected 😅
> 
> ```
> > JSON5.stringify(-45n)
> undefined
> > JSON.stringify(-45n)
> Uncaught TypeError: Do not know how to serialize a BigInt
>     at JSON.stringify (<anonymous>)
> ```
> 
> Would it be possible to delete the round tripping, use a different serialization library, or do you have any other suggestions?

Oh no, silent failures are so horrible! 

There is a quite a bit cloning, sometimes to protect an internal mutable rep. We also serialize/de-serialize pinned and human-annotated tests to/from disk. The serialization limitations and gotchas have been a persistent pain & would love to find a solution that works better ... hopefully without forking or writing our own. 😅

There are some suggestions on MDN for handling bigints with JSON that might work with JSON5 also---all of these are imperfect, but do any of these look potentially useful? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#use_within_json

**Comment by ZiyaoWei** [comment]
Thanks for the comments @cmumatt!

> Yes, signed vs. unsigned is a very old setting. I believe we only used it at the very beginning prior to adding ranges, so agree that it's superfluous at this point.

Would cleaning it up be a good thing to do or do you think it should be there as is?

> a solution that works better

I don't have any suggestions right now but as you know I spotted #301 too that's now two issues in a few months :sweat_smile so maybe @taisnguyen can think about it while he is working on the other issue?

EDIT: Maybe one of the following solutions can scale up to handling extra cases but might be a bit janky.

> do any of these look potentially useful?

Thanks and I think yes! After skimming through relevant sections I can think of multiple potential solutions right now:

- Use a string magic prefix maybe `bigint-[random bytes]-` and track all such magic values somewhere to avoid generating them during fuzzing and alert the users if one value happen to be a test case - extremely unlikely of course :-)
  - (+) Simple and should basically always work
  - (+) Kinda "scales" to other types - in quotes due to its jankiness, but I can imagine using a similar solution for other types and at least there could be a central place to track these.
  - (-) On the small chance it doesn't is super janky
  - (-) A bit of performance penalty but should be OK since the (de)serialization is probably quite expensive
- Go one step further and always serialize a type tag
  - (+) Should be straightforward and cover all cases?
  - (-) Performance might be a lot worse
  - (-) Basically writing our own library
- Serialize as numbers: if a number is integer and > 2^53-1 deserialize to bigint, otherwise deserialize to number, and if it turns out a bigint is expected convert to bigint at use sites
  - (+) Simple
  - (-) Works for bigint only
  - (?) I think the conversion from number to bigint at use sites is doable but I am not 100% sure
- Use https://www.npmjs.com/package/json-with-bigint which seems to be basically the same thing
  - (+) Somewhat popular a lot of downloads
  - (-) Works for bigint only
  - (-) Use functions provided there and not standard functions
- Serializes as strings; if it turns out a bigint is expected convert to bigint at use sites
  - (-) Similar tradeoffs as using numbers but feels jankier and maybe slower too?

Curious which do you prefer or do you have any other preferred solutions?

**Comment by cmumatt** [comment]
> > Yes, signed vs. unsigned is a very old setting. I believe we only used it at the very beginning prior to adding ranges, so agree that it's superfluous at this point.
> 
> Would cleaning it up be a good thing to do or do you think it should be there as is?

Deleting code is the among the highest callings---I'm for it!

> > a solution that works better
> 
> I don't have any suggestions right now but as you know I spotted [#301](https://github.com/nanofuzz/nanofuzz/issues/301) too that's now two issues in a few months :sweat_smile so maybe [@taisnguyen](https://github.com/taisnguyen) can think about it while he is working on the other issue?

For sure, it's really an endless pain. @taisnguyen showed me the approach he is working on for some of the non-bigint cases and wants to build out more tests prior to creating a PR. He's in the throes of finals, graduation, and about to move so those are taking a lot of his time presently. The approach he showed me wraps JSON5 and uses magic values similar to how `AiInputGenerator` works around some special cases particular to Zod's limitations.

> Thanks and I think yes! After skimming through relevant sections I can think of multiple potential solutions right now:
> 
>     * Use a string magic prefix maybe `bigint-[random bytes]-` and track all such magic values somewhere to avoid generating them during fuzzing and alert the users if one value happen to be a test case - extremely unlikely of course :-)
>       
>       * (+) Simple and should basically always work
>       * (+) Kinda "scales" to other types - in quotes due to its jankiness, but I can imagine using a similar solution for other types and at least there could be a central place to track these.
>       * (-) On the small chance it doesn't is super janky
>       * (-) A bit of performance penalty but should be OK since the (de)serialization is probably quite expensive
> 
>     * Go one step further and always serialize a type tag
>       
>       * (+) Should be straightforward and cover all cases?
>       * (-) Performance might be a lot worse
>       * (-) Basically writing our own library
> 
>     * Serialize as numbers: if a number is integer and > 2^53-1 deserialize to bigint, otherwise deserialize to number, and if it turns out a bigint is expected convert to bigint at use sites
>       
>       * (+) Simple
>       * (-) Works for bigint only
>       * (?) I think the conversion from number to bigint at use sites is doable but I am not 100% sure
> 
>     * Use https://www.npmjs.com/package/json-with-bigint which seems to be basically the same thing
>       
>       * (+) Somewhat popular a lot of downloads
>       * (-) Works for bigint only
>       * (-) Use functions provided there and not standard functions
> 
>     * Serializes as strings; if it turns out a bigint is expected convert to bigint at use sites
>       
>       * (-) Similar tradeoffs as using numbers but feels jankier and maybe slower too?
> 
> Curious which do you prefer or do you have any other preferred solutions?

Out of these, the "one step further" option looks most interesting if it can solve the problems. Hopefully the perf hit won't be too bad?

**Comment by ZiyaoWei** [comment]
I can try prototyping something and see if it works! Going to do the cleanup too. Curious is @taisnguyen's (congrats!) solution somewhere I can check out, maybe a branch? Don't wanna duplicate work and if we can unify things it might be better.

**Comment by cmumatt** [comment]
> I can try prototyping something and see if it works! Going to do the cleanup too. Curious is [@taisnguyen](https://github.com/taisnguyen)'s (congrats!) solution somewhere I can check out, maybe a branch? Don't wanna duplicate work and if we can unify things it might be better.

Yeah he has it in a fork somewhere---I've only seen it over Zoom so far. 🤓

**Comment by cmumatt** [comment]
JSONN serialization support for bigints landed in #421, which removes _that_ as a blocker for _this_ issue.

**Comment by ZiyaoWei** [comment]
Great, thanks! I’ll see if it I can figure out the rest over the weekend!
